seastar/src/submit_to.rs

```diff
@@ -0,0 +1,110 @@
+use crate as seastar;
+use ffi::*;
+use std::future::Future;
+
+use crate::cxx_async_local_future::IntoCxxAsyncLocalFuture;
+
+#[cxx::bridge]
+mod ffi {
+    #[namespace = "seastar_ffi"]
+    unsafe extern "C++" {
+        type VoidFuture = crate::cxx_async_futures::VoidFuture;
+    }
+
+    #[namespace = "seastar_ffi::submit_to"]
+    unsafe extern "C++" {
+        include!("seastar/src/submit_to.hh");
+
+        unsafe fn submit_to(
+            _shard_id: u32,
+            closure: *mut u8,
+            caller: unsafe fn(*mut u8) -> VoidFuture,
+        ) -> VoidFuture;
+    }
+}
+
+/// Convert the pointer to Box<Func> and then call it, consuming it in the process.
+fn fn_once_caller<Func>(raw_func: *mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe { Box::from_raw(raw_func as *mut Func)() }
+}
+
+/// A helper function. We need to be able to name the type of a named closure.
+/// There is no `typeof` in Rust, so this is achieved by using a dummy paramter.
+const fn get_fn_once_caller<Func>(_: &Func) -> fn(*mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_caller::<Func>
+}
+
+fn fn_once_dropper<Func>(raw_func: *mut u8)
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe {
+        let _ = Box::from_raw(raw_func as *mut Func);
+    }
+}
+
+const fn get_fn_once_dropper<Func>(_: &Func) -> fn(*mut u8)
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_dropper::<Func>
+}
+
+/// Runs a function `func` on a `shard_id` shard.
+///
+/// # Example
+///
+/// ```rust
+/// #[seastar::test]
+/// async fn submit_to_example() {
+///     let ret = submit_to(0, || future::ready(42)).await;
+///     assert!(ret.is_ok());
+///     assert_eq!(ret.unwrap(), 42);
+/// }
+pub async fn submit_to<Func, Fut, Ret>(shard_id: u32, func: Func) -> cxx_async::CxxAsyncResult<Ret>
+where
+    Func: FnOnce() -> Fut + Send + 'static,
+    Fut: Future<Output = Ret> + 'static,
+    Ret: Send + std::fmt::Debug + 'static,
+{
+    let (tx, rx) = futures::channel::oneshot::channel::<Ret>();
+
+    let closure = move || {
+        VoidFuture::infallible_local(async {
+            tx.send(func().await).unwrap();
+        })
+    };
+
+    let closure_caller = get_fn_once_caller(&closure);
+    let dropper = get_fn_once_dropper(&closure);
+    let boxed_closure = Box::into_raw(Box::new(closure)) as *mut u8;
+
+    unsafe {
+        match ffi::submit_to(shard_id, boxed_closure, closure_caller).await {
+            Ok(_) => Ok(rx.await.unwrap()),
+            Err(e) => {
+                dropper(boxed_closure);
+                Err(e)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I think it'd be good to have a test of nested `submit_to`s on top of the existing one.